### ISSUE TITLE
Entity with ToQuery throws InvalidOperationException in null check after Select

### ISSUE THREAD

**Issue by cyptus** [dotnet/efcore#25784]
When adding a null check in DbSet queries which are declared by "ToQuery" `ef core 3.1.18` throws with:

```
System.InvalidOperationException: 'The LINQ expression 'DbSet<Group>
    .Select(g => new GroupWrap{ Group = g }
    )
    .Where(g0 => g0.Group != null && g0.Group.GroupId == 1)' could not be translated. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to either AsEnumerable(), AsAsyncEnumerable(), ToList(), or ToListAsync(). See https://go.microsoft.com/fwlink/?linkid=2101038 for more information.'
``` 

DbContext:

```C#
        public DbSet<Group> Groups { get; set; }
        public DbSet<GroupWrap> GroupWraps { get; set; }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder
                .Entity<GroupWrap>()
                .ToQuery(() =>
                    from g in Groups
                    select new GroupWrap()
                    {
                        Group = g
                    }
                )
                .HasNoKey();

            base.OnModelCreating(modelBuilder);
        }
```

Entities:

```C#
    public class Group
    {
        [Key]
        public int GroupId { get; set; }
        public string Name { get; set; }
    }
    public class GroupWrap
    {
        [NotMapped]
        public Group Group { get; set; }
    }
```

Test:

```C#
            using (var db = new AppDbContext())
            {
                db.Database.EnsureDeleted();
                db.Database.EnsureCreated();

                db.Groups.Add(new Group()
                {
                    Name = "First"
                });

                db.SaveChanges();
            }

            using (var db = new AppDbContext())
            {
                //works
                var data1 = db.GroupWraps.Where(x => x.Group.GroupId == 1).ToList();

                //fails
                var data2 = db.GroupWraps.Where(x => x.Group != null && x.Group.GroupId == 1).ToList();
            }
```

The Wrapping-Class is just a demo, in real case it is filled with additional data from joins and subselects.

The same query against "Group" entity directly is working fine:

```C#

                var data = db.Groups.Select(x => new GroupWrap()
                    {
                        Group = x
                    })
                    .Where(x => x.Group != null && x.Group.GroupId == 1)
                    .ToList();
```

**Comment by cyptus** [comment]
[ConsoleAppEfCoreTest.zip](https://github.com/dotnet/efcore/files/7078006/ConsoleAppEfCoreTest.zip)

Demo Project attached

**Comment by ajcvickers** [comment]
@cyptus The `ToQuery` API is obsoleted in version 5.0 because of issues with it's design. Making changes to fix issues in the 3.1 implementation [doesn't meet the bar for a patch release](https://docs.microsoft.com/en-us/ef/core/what-is-new/release-planning). Consider updating to EF Core 5.0 and using `ToSqlQuery`, as described here: https://docs.microsoft.com/en-us/ef/core/what-is-new/ef-core-5.0/breaking-changes#defining-query-is-replaced-with-provider-specific-methods
